### PR TITLE
fix: can't sync temporal flag on virtual table

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

Currently, the virtual table is unable to sync `is temporal` flag. This is due to the `SupersetResultSet` response `is_date` instead of `is_dttm`. We should keep the physical table and virtual table have the same interface. 

In this PR also introduce a new type `superset.superset_typing.ResultSetColumnType` to guarantee type checking.

I will address a separate PR to finish `PhysicalColumnType`.

#### Before

https://user-images.githubusercontent.com/2016594/160112782-20110704-6a84-4004-b8d2-3c889871b2a1.mov

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by jinghua-qa** [comment]
I tested the PR locally, is temporal can be sync as expected, but when i save the changes, i saw error msg

https://user-images.githubusercontent.com/81597121/160926244-a9bcd466-3a04-474f-b76f-ae32421a4174.mov

**Comment by zhaoyongjie** [comment]
> I tested the PR locally, is temporal can be sync as expected, but when i save the changes, i saw error msg
> Screen.Recording.2022-03-30.at.1.34.07.PM.mov

This is an unrelated issue with current PR. I have opened a separated [PR](https://github.com/apache/superset/pull/19452) to fix it.

**Review by jinghua-qa — APPROVED** [review]
Work as expected, LGTM

File: superset/result_set.py [hunk 1]
```diff
@@ -210,17 +210,17 @@ def size(self) -> int:
...
 
         columns = []
         for col in self.table.schema:
             db_type_str = self.data_type(col.name, col.type)
-            column = {
+            column: ResultSetColumnType = {
                 "name": col.name,
                 "type": db_type_str,
-                "is_date": self.is_temporal(db_type_str),
+                "is_dttm": self.is_temporal(db_type_str),
```
Comment (zhaoyongjie) [inline superset/result_set.py:223]: we should keep a consistent interface between the physical dataset and the virtual dataset.

**Review by villebro — APPROVED** [review]
Love it, thanks for this fix + nice cleanup! In the long run I feel we should start moving types from `superset/superset_typing.py` into the packages where they are most relevant (see e.g. `superset/key_value/types.py`), but that can be done in a follow up (e.g. the one planned for physical columns)

**Comment by zhaoyongjie** [comment]
> Love it, thanks for this fix + nice cleanup! In the long run I feel we should start moving types from `superset/superset_typing.py` into the packages where they are most relevant (see e.g. `superset/key_value/types.py`), but that can be done in a follow up (e.g. the one planned for physical columns)

sounds good! I will follow this pattern in future.

**Comment by sadpandajoe** [comment]
🏷️ preset:2022.13

**Pull request merged**